src/main/java/com/snowflake/kafka/connector/internal/streaming/schemaevolution/iceberg/IcebergSchemaEvolutionService.java

```diff
@@ -17,48 +20,163 @@ public class IcebergSchemaEvolutionService implements SchemaEvolutionService {
+
+    Set<String> columnsToEvolve = extractColumnNames(targetItems);
+
     // Add columns if needed, ignore any exceptions since other task might be succeeded
-    if (!columnsToAdd.isEmpty()) {
-      LOGGER.debug("Adding columns to iceberg table: {} columns: {}", tableName, columnsToAdd);
-      TableSchema tableSchema =
-          tableSchemaResolver.resolveTableSchemaFromRecord(record, columnsToAdd);
-      try {
-        conn.appendColumnsToIcebergTable(tableName, tableSchema.getColumnInfos());
-      } catch (SnowflakeKafkaConnectorException e) {
-        LOGGER.warn(
-            String.format(
-                "Failure altering iceberg table to add column: %s, this could happen when multiple"
-                    + " partitions try to alter the table at the same time and the warning could be"
-                    + " ignored",
-                tableName),
-            e);
-      }
+    if (!columnsToEvolve.isEmpty()) {
+      LOGGER.debug("Adding columns to iceberg table: {} columns: {}", tableName, columnsToEvolve);
+      // some of the column might already exist, and we will modify them, not create
+      List<IcebergColumnTree> alreadyExistingColumns =
+          icebergTableSchemaResolver.resolveIcebergSchemaFromChannel(
+              existingSchema, columnsToEvolve);
+
+      List<IcebergColumnTree> modifiedOrAddedColumns =
+          icebergTableSchemaResolver.resolveIcebergSchemaFromRecord(record, columnsToEvolve);
+
+      List<IcebergColumnTree> columnsToAdd =
+          distinguishColumnsToAdd(alreadyExistingColumns, modifiedOrAddedColumns);
+
+      List<IcebergColumnTree> columnsToModify =
+          distinguishColumnsToModify(alreadyExistingColumns, modifiedOrAddedColumns);
+
+      alterAddColumns(tableName, columnsToAdd);
+
+      alterDataType(tableName, alreadyExistingColumns, columnsToModify);
+    }
+  }
+
+  /**
+   * Get only column names, ignore nested field names, remove double quotes.
+   *
+   * <p>example: TEST_STRUCT.field1 -> TEST_STRUCT
+   */
+  private Set<String> extractColumnNames(SchemaEvolutionTargetItems targetItems) {
+    return targetItems.getColumnsToAdd().stream()
+        .map(this::removeNestedFieldNames)
+        .map(this::removeDoubleQuotes)
+        .collect(Collectors.toSet());
+  }
+
+  private String removeNestedFieldNames(String columnNameWithNestedNames) {
+    return columnNameWithNestedNames.split("\\.")[0];
+  }
+
+  private String removeDoubleQuotes(String columnName) {
+    return columnName.replaceAll("\"", "");
+  }
+
+  /** Columns that are not present in a current schema are to be added */
+  private List<IcebergColumnTree> distinguishColumnsToAdd(
+      List<IcebergColumnTree> alreadyExistingColumns,
+      List<IcebergColumnTree> modifiedOrAddedColumns) {
+    return modifiedOrAddedColumns.stream()
+        .filter(
+            modifiedOrAddedColumn ->
+                alreadyExistingColumns.stream()
+                    .noneMatch(
```

Comment:
It would be easier to convert `alreadyExistingColumns` from List to Set.

Reply:
I can use Set everywhere instead of a list.

Reply:
Good point. Is there any problem with implementing these methods or the final code doesn't look cleaner?

Reply:
If we implement only for columnName - there is not a problem. Using it for both doesn't make sense. 
I don't think using Set will change anything in a logic. It may (shouldn't) sneakily replace a column somewhere... I can change it, never say never.